File: src/astro/surpalaces.rs
use crate::data::stars::StarKey;
use crate::data::types::Mutagen;
use crate::models::palace::PalaceData;

/// 三方四正宫位
pub struct SurroundedPalaces<'a> {
    /// 本宫
    pub target: &'a PalaceData,
    /// 对宫 (+6)
    pub opposite: &'a PalaceData,
    /// 财帛位 (+8, modular)
    pub wealth: &'a PalaceData,
    /// 官禄位 (+4)
    pub career: &'a PalaceData,
}

impl<'a> SurroundedPalaces<'a> {
    /// 三方四正是否包含所有指定星耀
    pub fn have(&self, stars: &[StarKey]) -> bool {
        let all_keys = self.all_star_keys();
        stars.iter().all(|s| all_keys.contains(s))
    }

    /// 三方四正是否不含所有指定星耀
    pub fn not_have(&self, stars: &[StarKey]) -> bool {
        let all_keys = self.all_star_keys();
        stars.iter().all(|s| !all_keys.contains(s))
    }

    /// 三方四正是否包含任一指定星耀
    pub fn have_one_of(&self, stars: &[StarKey]) -> bool {
        let all_keys = self.all_star_keys();
        stars.iter().any(|s| all_keys.contains(s))
    }

    /// 三方四正是否有指定四化
    pub fn have_mutagen(&self, mutagen: Mutagen) -> bool {
        self.target.has_mutagen(mutagen)
            || self.opposite.has_mutagen(mutagen)
            || self.wealth.has_mutagen(mutagen)
            || self.career.has_mutagen(mutagen)
    }

    /// 三方四正是否没有指定四化
    pub fn not_have_mutagen(&self, mutagen: Mutagen) -> bool {
        !self.have_mutagen(mutagen)
    }

    /// 收集三方四正所有星耀的 StarKey
    fn all_star_keys(&self) -> Vec<StarKey> {
        let palaces = [self.target, self.opposite, self.wealth, self.career];
        let mut keys = Vec::new();
        for p in &palaces {
            for s in &p.major_stars {
                keys.push(s.key);
            }
            for s in &p.minor_stars {
                keys.push(s.key);
            }
            for s in &p.adjective_stars {
                keys.push(s.key);
            }
        }
        keys
    }
}

#[cfg(test)]
mod tests {
    use crate::astro::builder::by_solar;
    use crate::data::stars::StarKey;
    use crate::data::types::*;

    fn make_astrolabe() -> crate::models::astrolabe::Astrolabe {
        by_solar(
            "2000-8-16",
            2,
            Gender::Female,
            true,
            Language::ZhCN,
            Config::default(),
        )
        .unwrap()
    }

    #[test]
    fn test_surrounded_palaces_structure() {
        let a = make_astrolabe();
        let sp = a.surrounded_palaces(0);
        // target is palace 0
        assert_eq!(sp.target.index, 0);
        // opposite is palace 6
        assert_eq!(sp.opposite.index, 6);
        // career is palace 4
        assert_eq!(sp.career.index, 4);
        // wealth is palace 8
        assert_eq!(sp.wealth.index, 8);
    }

    #[test]
    fn test_surrounded_palaces_wrapping() {
        let a = make_astrolabe();
        let sp = a.surrounded_palaces(10);
        assert_eq!(sp.target.index, 10);
        // opposite: (10+6) % 12 = 4
        assert_eq!(sp.opposite.index, 4);
        // career: (10+4) % 12 = 2
        assert_eq!(sp.career.index, 2);
        // wealth: (10+8) % 12 = 6
        assert_eq!(sp.wealth.index, 6);
    }

    #[test]
    fn test_surrounded_palaces_have() {
        let a = make_astrolabe();
        // The surrounded palaces of palace 0 cover 4 palaces,
        // so they should have many stars combined
        let sp = a.surrounded_palaces(0);
        // At least one major star should be present across 4 palaces
        assert!(sp.have_one_of(&[
            StarKey::ZiweiMaj,
            StarKey::TianjiMaj,
            StarKey::TaiyangMaj,
            StarKey::TianfuMaj,
        ]));
    }

    #[test]
    fn test_surrounded_palaces_not_have() {
        let a = make_astrolabe();
        let sp = a.surrounded_palaces(0);
        // Collect all star keys in the surrounded palaces
        // Then pick a star NOT present to test not_have
        let all_majors = [
            StarKey::ZiweiMaj,
            StarKey::TianjiMaj,
            StarKey::TaiyangMaj,
            StarKey::WuquMaj,
            StarKey::TiantongMaj,
            StarKey::LianzhenMaj,
            StarKey::TianfuMaj,
            StarKey::TaiyinMaj,
            StarKey::TanlangMaj,
            StarKey::JumenMaj,
            StarKey::TianxiangMaj,
            StarKey::TianliangMaj,
            StarKey::QishaMaj,
            StarKey::PojunMaj,
        ];
        let missing: Vec<StarKey> = all_majors
            .iter()
            .filter(|k| !sp.have(&[**k]))
            .copied()
            .collect();
        if !missing.is_empty() {
            assert!(sp.not_have(&missing));
        }
    }

    #[test]
    fn test_surrounded_palaces_mutagen() {
        let a = make_astrolabe();
        // Check across all 12 positions — at least one should have Lu mutagen
        let any_has_lu = (0..12).any(|i| a.surrounded_palaces(i).have_mutagen(Mutagen::Lu));
        assert!(any_has_lu);
        let any_has_ji = (0..12).any(|i| a.surrounded_palaces(i).have_mutagen(Mutagen::Ji));
        assert!(any_has_ji);
    }

    #[test]
    fn test_surrounded_palaces_not_have_mutagen() {
        let a = make_astrolabe();
        let sp = a.surrounded_palaces(0);
        // not_have_mutagen is the inverse of have_mutagen
        for m in &[Mutagen::Lu, Mutagen::Quan, Mutagen::Ke, Mutagen::Ji] {
            assert_eq!(sp.not_have_mutagen(*m), !sp.have_mutagen(*m));
        }
    }
}

File: python/x_iztro/surpalaces.py
"""
三方四正。

紫微斗数不单看一宫：本宫、对宫、财帛位、官禄位四宫合看，
`SurroundedPalaces` 把这四宫打包，判断方法对四宫取并集。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from x_iztro.enums import Mutagen
from x_iztro.palace import Palace
from x_iztro.star_object import _star_identifiers

if TYPE_CHECKING:
    from x_iztro.knowledge import KnowledgePack


@dataclass(frozen=True, slots=True)
class SurroundedPalaces:
    """三方四正宫位"""

    target: Palace
    """本宫"""

    opposite: Palace
    """对宫"""

    wealth: Palace
    """财帛位（三方）"""

    career: Palace
    """官禄位（三方）"""

    def to_text(self, *, knowledge: bool | KnowledgePack | None = None) -> str:
        """
        三方四正的语义化文本：本宫、对宫、财帛位、官禄位四宫合看的完整描述。

        从本宫所属星盘的排盘上下文（含重排起点）无状态再发起计算，
        文本按排盘语言输出。

        Args:
            knowledge: 释义材料（True 取排盘语言的内嵌包，或给 KnowledgePack）；
                给出时追加四宫星耀的释义节

        Raises:
            ValueError: 本宫脱离星盘单独构造，无排盘上下文可转发
            IztroError: `knowledge=True` 而排盘语言没有内嵌包（目前只有 zh-CN）
        """
        astrolabe = self.target.astrolabe()
        if astrolabe is None:
            raise ValueError(
                "to_text 需要所属星盘：请从 Astrolabe.surrounded_palaces 获取三方四正"
            )
        return astrolabe._context_query(
            "surroundedPalacesToText", palace_index=self.target.index, knowledge=knowledge
        )

    def have(self, stars: list[str]) -> bool:
        """判断三方四正是否包含指定的 **所有** 星耀（接受星耀枚举或当前语言的星名）"""
        identifiers = self._all_star_identifiers()
        return all(s in identifiers for s in stars)

    def not_have(self, stars: list[str]) -> bool:
        """判断三方四正是否 **不** 包含指定的所有星耀"""
        identifiers = self._all_star_identifiers()
        return all(s not in identifiers for s in stars)

    def have_one_of(self, stars: list[str]) -> bool:
        """判断三方四正是否包含指定星耀中的 **至少一颗**"""
        identifiers = self._all_star_identifiers()
        return any(s in identifiers for s in stars)

    def have_mutagen(self, mutagen: Mutagen) -> bool:
        """判断三方四正中是否有指定四化"""
        return any(p.has_mutagen(mutagen) for p in self._all_palaces())

    def not_have_mutagen(self, mutagen: Mutagen) -> bool:
        """判断三方四正中是否没有指定四化"""
        return not self.have_mutagen(mutagen)

    def _all_palaces(self) -> list[Palace]:
        return [self.target, self.opposite, self.wealth, self.career]

    def _all_star_identifiers(self) -> set[str]:
        out: set[str] = set()
        for p in self._all_palaces():
            out |= _star_identifiers(
                p.major_stars + p.minor_stars + p.adjective_stars
            )
        return out
